cms/core/models/base.py

```diff
@@ -79,3 +86,35 @@ def has_ons_embed(self) -> bool:
                 return False
 
         return False
+
+
+class BaseSiteSetting(WagtailBaseSiteSetting):
+    """A customized site setting.
+
+    - Use default values in external environment if an instance doesn't exist.
+    - Use `.get` to ensure the read connection is used.
+    """
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def for_site(cls, site: Optional["Site"]) -> Self:
+        """Get or create an instance of this setting for the site."""
+        if site is None:
+            raise cls.DoesNotExist(f"{cls} does not exist for site None.")
+
+        queryset = cls.base_queryset()
+
+        try:
+            # Explicitly call `.get` first to ensure the
+            # read connection is used.
+            return cast(Self, queryset.get(site=site))
+        except cls.DoesNotExist:
+            if settings.IS_EXTERNAL_ENV:
+                # In the external env, the database connection is read only,
+                # so just use the default values if the instance doesn't exist.
+                return cls(site=site)
+
+            instance, _created = queryset.get_or_create(site=site)
```

Comment:
I was wondering why Django does not handle this for us by using read for `get` first, then `write`? But I saw this comment in the `get_or_create` method.
```py
# The get() needs to be targeted at the write database in order
# to avoid potential transaction consistency problems.
```
Could this not be a problem for us (edge case)?

Reply:
No, for us this is fine. `get_or_create` is itself internally consistent. If it doesn't exist, it attempts to create it. If it did infact exist, just read it again.

The reason for this override is to ensure that even the first `.get` uses the read connection, which `get_or_create` doesn't. Since 99% of the times this is run, there'll be an instance, it's worth the extra bit of code to ensure the replicas are used.